
# load packages----
suppressPackageStartupMessages({  
  library(rhdf5) 
  library(tidyverse) 
  library(tximport) 
  library(ensembldb) 
  library(EnsDb.Hsapiens.v86) 
  })

# you can easily create file paths to the abundance files generated by Kallisto using the 'file.path' function
path <- file.path(paste0("../data/raw_data/aligned/", phenodataF$Run), "abundance.tsv") # set file paths to your mapped data
# now check to make sure this path is correct by seeing if the files exist
all(file.exists(path)) 
which(!file.exists(path))  #if you get false from line above, check which one is false.



# get annotations using organism-specific package ----
supportedFilters(EnsDb.Hsapiens.v86)
Tx <- transcripts(EnsDb.Hsapiens.v86, columns=c("tx_id", "gene_name", "entrezid")) #ensembldb package
Tx <- as_tibble(Tx)
#need to change first column name to 'target_id'
Tx <- dplyr::rename(Tx, target_id = tx_id)
TxOG <- Tx
TxOG <- TxOG %>% rename_with(~"gene", gene_name)

#transcrip ID needs to be the first column in the dataframe
Tx <- dplyr::select(Tx, "target_id", "gene_name")


# import Kallisto transcript counts into R using Tximport ----
# copy the abundance files to the working directory and rename so that each sample has a unique name
Txi_gene <- tximport(path, 
                     type = "kallisto", 
                     tx2gene = Tx, 
                     txOut = FALSE, #How does the result change if this =FALSE vs =TRUE?
                     countsFromAbundance = "lengthScaledTPM",
                     ignoreTxVersion = TRUE)

#take a look at the type of object you just created
class(Txi_gene)
names(Txi_gene)

print("Step 1 complete!")


# if you want to write your counts or TPM to a file on your harddrive
# if you exported transcript level data and want to append your gene symbols to the data frame
# Txi_trans <- as_tibble(Txi_trans$counts, rownames = "target_id")
# Txi_trans <- left_join(Txi_trans, Tx)

